inverse dynamics controlled kuka arm simulation

siyuanfeng-tri opened the pull request:
0. change the qp controller / humanoid status to support with fixed base robot.
1. added a kuka arm controlled with inverse dynamics demo
2. changed the qp inverse dynamics system to be discrete time.
3. sub class plan eval and joint level controller systems to support different robots. 

+@sammy-tri for feature review, thanks.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4992)
<!-- Reviewable:end -->

sammy-tri commented:
+(status: curate commits before merging) 

Review in progress, this might take me a bit.

---

Reviewed 16 of 40 files at r1.
Review status: 16 of 40 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/examples/QPInverseDynamicsForHumanoids/CMakeLists.txt, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbuS7ZtoPXg1jXJR-pT:-KbuS7ZtoPXg1jXJR-pU:bn5k1yd) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/CMakeLists.txt#L30)):*
> ```CMake
> endif()
> ```

I'm not sure why we're adding extra vertical whitespace at the end of the CMakeLists.txt file here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KbuQkjJkLdsOdoEjlJp:bd6kpz7)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 5 of 40 files at r1.
Review status: 21 of 40 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/atlas_joint_level_controller_system.cc, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-Kbv8jkEzRtUA85ZUA8G:-Kbv8jkEzRtUA85ZUA8H:b-qvx553) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/atlas_joint_level_controller_system.cc#L54)):*
> ```cpp
>   // The torques have already been computed and set in
>   // JointLevelControllerSystem::DoCalcOutput()
>   auto act_torques = GetMutableOutputVector(output, output_port_index_torque_);
> ```

Why are we using the mutable version here?

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 76 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvCs514HeHEFLyWRZF:-KbvCs514HeHEFLyWRZG:b-qzzd8c) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L76)):*
> ```objc
>  protected:
>   template <typename PlanType>
>   PlanType& get_mutable_plan(systems::State<double>* state) const {
> ```

variants of get_mutable_* typically return pointers instead of non-const references (looking in context.h for example)

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 94 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvEy91XHk3rTDiIxOe:-KbvEy91XHk3rTDiIxOf:b-6ynmd2) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L94)):*
> ```objc
>   param_parsers::ParamSet paramset_;
> 
>   int input_port_index_humanoid_status_;
> ```

These may be required to have inline initializers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 40 files at r1.
Review status: 23 of 40 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/make_demo_plan.h, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvGgzeByiG-lrundkr:-KbvGgzeByiG-lrundks:b-gdvojz) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/kuka_iiwa_arm/controlled_kuka/make_demo_plan.h#L12)):*
> ```objc
> namespace kuka_iiwa_arm {
> 
> std::unique_ptr<PiecewisePolynomialTrajectory>
> ```

This could use some documentation (what does the "path" argument do?  What are the properties of the returned trajectory?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 15 of 43 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/make_demo_plan.h, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvGgzeByiG-lrundkr:-KbvM7OTnEsxhg_QUWJ2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/kuka_iiwa_arm/controlled_kuka/make_demo_plan.h#L12)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This could use some documentation (what does the "path" argument do?  What are the properties of the returned trajectory?)
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/CMakeLists.txt, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbuS7ZtoPXg1jXJR-pT:-KbvMFBo9oPRcCC4zBkK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/CMakeLists.txt#L30)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'm not sure why we're adding extra vertical whitespace at the end of the CMakeLists.txt file here
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/atlas_joint_level_controller_system.cc, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-Kbv8jkEzRtUA85ZUA8G:-KbvNWNd5svyRjXXxT_Y:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/atlas_joint_level_controller_system.cc#L54)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Why are we using the mutable version here?
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 76 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvCs514HeHEFLyWRZF:-KbvNs_2esfl81QzK3YP:b1gpt7j) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L76)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

variants of get_mutable_* typically return pointers instead of non-const references (looking in context.h for example)
</blockquote></details>

hmm, i copied these from abstract state's mutable getters, it turns out they do &. it's kind of silly to switch between * and & on my end. I am going to leave it as is.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 94 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvEy91XHk3rTDiIxOe:-KbvOjZli5skC4FoUmeD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L94)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

These may be required to have inline initializers.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 12 of 20 files at r2.
Review status: 27 of 43 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 76 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbvCs514HeHEFLyWRZF:-KbvcvUG7HngQoqCU0bj:b-83scg6) ([raw file](https://github.com/robotlocomotion/drake/blob/43dc162b3c1ada123bfc4cc6ea25f1287524f8fd/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L76)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

hmm, i copied these from abstract state's mutable getters, it turns out they do &. it's kind of silly to switch between * and & on my end. I am going to leave it as is.
</blockquote></details>

Huh, so it does.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 8 of 40 files at r1, 8 of 20 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbzP6tsfO6yOjhSIQcb:-KbzP6ttmdLobc6-M5xi:b-87feee) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1eb1487a1087f2591d717cee64f2aca84b4d1f/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L87)):*
> ```objc
>   }
> 
>   const RigidBodyTree<double>& robot_;
> ```

All data members are required to be private: https://google.github.io/styleguide/cppguide.html#Access_Control

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/humanoid_plan_eval_system.cc, line 56 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbzPRzrmS0pzWCRJuLm:-KbzPRzshLwP5Guha3gb:bk29803) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1eb1487a1087f2591d717cee64f2aca84b4d1f/drake/examples/QPInverseDynamicsForHumanoids/system/humanoid_plan_eval_system.cc#L56)):*
> ```cpp
>   QpInput& qp_input = get_mutable_qp_input(state);
>   qp_input = paramset_.MakeQpInput({"feet"},            /* contacts */
>                                    {"pelvis", "torso"}, /* tracekd bodies */
> ```

typo "tracked"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Completed another review pass.

---

Reviewed 15 of 17 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/CMakeLists.txt, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY_1PiSEeBJ0k7Pc58:-KcY_1PjuJ6nJX5kTGM_:b-ljr0hx) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/examples/kuka_iiwa_arm/controlled_kuka/CMakeLists.txt#L1)):*
> ```CMake
> add_definitions(-g)
> ```

Is the leftover debugging code?

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_inverse_dynamics_demo.cc, line 59 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcYcOli0CmXBDUXETxX:-KcYcOlj6Px9klG4sL-6:b5krz29) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_inverse_dynamics_demo.cc#L59)):*
> ```cpp
>       std::make_unique<RigidBodyTree<double>>();
>   drake::parsers::urdf::AddModelInstanceFromUrdfFile(
>       drake::GetDrakePath() +
> ```

BTW it might be less brittle to use model_path here too

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-KcY__YJenX7uUE1uWvZ:b-afwa0j) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
> ```objc
> #pragma once
> ```

Would it make more sense to add this functionality to PiecewisePolynomialTrajectory and use TrajectorySource instead of creating a new source system?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcYcYsH9stsY6RtyLyB:b-y093ec)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sammy-tri i am lumping some of the sprint code in this pr. so it's still WIP. my apologies. should make the sprint pr easier. 

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbzP6tsfO6yOjhSIQcb:-KcUm__PSbMhU9DnBVzx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1eb1487a1087f2591d717cee64f2aca84b4d1f/drake/examples/QPInverseDynamicsForHumanoids/system/discrete_time_plan_eval_system.h#L87)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

All data members are required to be private: https://google.github.io/styleguide/cppguide.html#Access_Control
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcYdlw2qPd4JMbLaU5F:b-camt1k)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sammy-tri i updated some comments, and is ready for review. thanks!

---

Reviewed 1 of 44 files at r1, 1 of 18 files at r3, 27 of 34 files at r4.
Review status: 40 of 54 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/CMakeLists.txt, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY_1PiSEeBJ0k7Pc58:-KcZP2Fv4xfLAASXCrUc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/examples/kuka_iiwa_arm/controlled_kuka/CMakeLists.txt#L1)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Is the leftover debugging code?
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_inverse_dynamics_demo.cc, line 59 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcYcOli0CmXBDUXETxX:-KcZPIsvdd1wDm28s0fr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/examples/kuka_iiwa_arm/controlled_kuka/kuka_inverse_dynamics_demo.cc#L59)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW it might be less brittle to use model_path here too
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/humanoid_plan_eval_system.cc, line 56 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KbzPRzrmS0pzWCRJuLm:-KcZPgHQIOz9z4X6H29r:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1eb1487a1087f2591d717cee64f2aca84b4d1f/drake/examples/QPInverseDynamicsForHumanoids/system/humanoid_plan_eval_system.cc#L56)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

typo "tracked"
</blockquote></details>

Done.

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-KcZPk83G2Cof_wqIp64:b5r7lfp) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Would it make more sense to add this functionality to PiecewisePolynomialTrajectory and use TrajectorySource instead of creating a new source system?
</blockquote></details>

the stupid thing is that trajectory source is supposed to be any trajectory, so presumably you can't take derivatives... 
do we use Trajectory other than a pp? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcZVoEtM7gb6pF_c2Gx:bsh0z9v)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please

sammy-tri commented:
Completed another pass.  The discussion about PiecewisePolynomialSource is still outstanding, but it might be time for platform review anyway.

---

Reviewed 1 of 18 files at r3, 21 of 34 files at r4, 1 of 2 files at r5, 12 of 12 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsEBX43HoJc4ZWVoan:-KcsEBX6fM4E1rnPrP0b:btm2ovz) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD#L2)):*
> ```Python
> # -*- python -*-
> # This file contains rules for Bazel; see drake/doc/bazel.rst.
> ```

BTW is it feasible to add the existing targets from CMakeLists.txt to this file?  Or maybe split up this directory into two?  I'm not sure I know the specific resolution I'm looking for, but it seems weird to me that this directory has an almost completely disjoint set of outputs based on the build system selected.

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/BUILD, line 96 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsFSjT01Jowsst7kA-:-KcsFSjT01Jowsst7kA0:b-h4jsip) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/QPInverseDynamicsForHumanoids/system/BUILD#L96)):*
> ```Python
> 
> drake_cc_library(
>     name = "id_contorlled_iiwa",
> ```

typo of controlled?

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-KcsGRo45PsOdPWoNMTj:bl9acva) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

the stupid thing is that trajectory source is supposed to be any trajectory, so presumably you can't take derivatives... 
do we use Trajectory other than a pp? 
</blockquote></details>

My thought was that adding the derivative order to PiecewisePolynomialTrajectory would change the number of rows in the output trajectory to include the derivatives, much like the size of this output port gets larger based on the order.

We don't actually specify anywhere that a trajectory is anything more than a time-varying matrix of doubles..   and we already have cases where that can be just positions or positions + velocities.

Pinging @lucy-tri and @russtedrake to see if either on has philosophical thoughts on if this type of thing should be using the Trajectory representation or if we want this alternate path to turning a PiecewisePolynomial into a SingleOutputVectorSource.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcsIxtP6rRER9F6re8d:b-dram87)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsEBX43HoJc4ZWVoan:-Kcsknc07hfSuyi8ADTR:b-q4zv76) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD#L2)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW is it feasible to add the existing targets from CMakeLists.txt to this file?  Or maybe split up this directory into two?  I'm not sure I know the specific resolution I'm looking for, but it seems weird to me that this directory has an almost completely disjoint set of outputs based on the build system selected.
</blockquote></details>

i can promise to port the cmake stuff to bazel, in the next pr, if that's your concern. 

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-Kcsl9EwID3YoadJxBFf:b-l20uwd) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

My thought was that adding the derivative order to PiecewisePolynomialTrajectory would change the number of rows in the output trajectory to include the derivatives, much like the size of this output port gets larger based on the order.

We don't actually specify anywhere that a trajectory is anything more than a time-varying matrix of doubles..   and we already have cases where that can be just positions or positions + velocities.

Pinging @lucy-tri and @russtedrake to see if either on has philosophical thoughts on if this type of thing should be using the Trajectory representation or if we want this alternate path to turning a PiecewisePolynomial into a SingleOutputVectorSource.

</blockquote></details>

My thought was that adding the derivative order to PiecewisePolynomialTrajectory would change the number of rows in the output trajectory to include the derivatives, much like the size of this output port gets larger based on the order.
- yes, you are correct. I can make it default to 0 order for derivatives to match the old behavior.

I would argue in the other direction that we do not have another form of Trajectory other than PP trajectory. And in the general case, we can represent  dense trajectories with a constant or linear pp. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsEBX43HoJc4ZWVoan:-Kct51DIdeShHsId5gid:bpfjlp3) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD#L2)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i can promise to port the cmake stuff to bazel, in the next pr, if that's your concern. 
</blockquote></details>

I think that would be an adequate solution...   until a cmake user needs this code to compile, but since bazel is the future, I won't claim you need to update CMakeLists.txt for the new code.

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-Kct5PdlMzv1-lc5mKjH:b-e794td) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

My thought was that adding the derivative order to PiecewisePolynomialTrajectory would change the number of rows in the output trajectory to include the derivatives, much like the size of this output port gets larger based on the order.
- yes, you are correct. I can make it default to 0 order for derivatives to match the old behavior.

I would argue in the other direction that we do not have another form of Trajectory other than PP trajectory. And in the general case, we can represent  dense trajectories with a constant or linear pp. 
</blockquote></details>

We don't yet have another form aside from PiecewisePolynomialTrajectory.  IIRC we did in MATLAB and there were plans to add more forms in the C++ Drake as of a couple of months ago, though clearly progress has been stalled (hence my ping to Lucy/Russ above as they were involved in those discussions).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-Kcwti55TTyoWJWMzCPN:bem4f2s) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

We don't yet have another form aside from PiecewisePolynomialTrajectory.  IIRC we did in MATLAB and there were plans to add more forms in the C++ Drake as of a couple of months ago, though clearly progress has been stalled (hence my ping to Lucy/Russ above as they were involved in those discussions).
</blockquote></details>

@sammy-tri i am happy to move that thing to the local directory. or even into the .cc file for the demo. i'd rather get this pr in first and talk about it later. what do you think? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 3 of 44 files at r1, 1 of 34 files at r4, 4 of 12 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-KcxIlGwtiLiNcups4Md:b-lkn84g) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
> ```objc
>   Eigen::Index cols() const override { return pp_.cols(); }
> 
>   const PiecewisePolynomial<double>& get_piecewise_polynomial() const {
> ```

In general, it's not that great an idea to create a setter to private member data and just return that.
Also, this method should have documentation.

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-KcxTSDEWg3dOhIbLw8d:b-69kup1) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

@sammy-tri i am happy to move that thing to the local directory. or even into the .cc file for the demo. i'd rather get this pr in first and talk about it later. what do you think? 
</blockquote></details>

It makes more sense to me to add this functionality to PiecewisePolynomialTrajectory, as Sam suggested. I guess the other question is whether the functionality should be added to the Trajectory class instead? Thus the questions of what other kinds of trajectories could their be?  Take a look at the Matlab Trajectory class http://drake.mit.edu/doxygen_matlab/class_trajectory.html, where you can see for example, ConstantTrajectory, HybridTrajectory, and others. In any case, if you think there might a problem taking derivatives on some kind of trajectory that doesn't allow it, then the only derivative order allowed for that subclass of Trajectory would be order 0.  Am I understanding all the issues?  

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 22 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxLqrUadMy3x_wYeFp:-KcxLqrV6b1Y6-5Q4R0r:bbw29r7) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/systems/primitives/piecewise_polynomial_source.h#L22)):*
> @p output_derivative_order.

Delete this line: you've listed the param below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
retracted the pp source from systems/primitive, moved it to the demo code for inverse dynamics controlled iiwa. 

---

Review status: 39 of 43 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-KcxiN_-QlHH-d9lNovn:b-wyfxr4) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

In general, it's not that great an idea to create a setter to private member data and just return that.
Also, this method should have documentation.
</blockquote></details>

it's a getter. 

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsEBX43HoJc4ZWVoan:-Kcxh7GLdA1j_uIHvbDE:b-juhhb9) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD#L2)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I think that would be an adequate solution...   until a cmake user needs this code to compile, but since bazel is the future, I won't claim you need to update CMakeLists.txt for the new code.
</blockquote></details>

for cmake, i will just make sure that the shared stuff i added is updated accordingly. 

---

*[drake/examples/QPInverseDynamicsForHumanoids/system/BUILD, line 96 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsFSjT01Jowsst7kA-:-KcxhaeC3Uq2J4dr8voP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/QPInverseDynamicsForHumanoids/system/BUILD#L96)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

typo of controlled?
</blockquote></details>

Done.

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-Kcxip1KglPZFD7-fsec:bo10018) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

It makes more sense to me to add this functionality to PiecewisePolynomialTrajectory, as Sam suggested. I guess the other question is whether the functionality should be added to the Trajectory class instead? Thus the questions of what other kinds of trajectories could their be?  Take a look at the Matlab Trajectory class http://drake.mit.edu/doxygen_matlab/class_trajectory.html, where you can see for example, ConstantTrajectory, HybridTrajectory, and others. In any case, if you think there might a problem taking derivatives on some kind of trajectory that doesn't allow it, then the only derivative order allowed for that subclass of Trajectory would be order 0.  Am I understanding all the issues?  
</blockquote></details>

i moved all that stuff to the actual demo code. i am leaving all the primitives as is. 

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 22 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxLqrUadMy3x_wYeFp:-Kcxhxb7bOHw9CqfaBMR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/systems/primitives/piecewise_polynomial_source.h#L22)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

> @p output_derivative_order.

Delete this line: you've listed the param below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-Kcxj6AXSDXxARwQU7rU:b-2iaj1e)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Moving the code to the demo code sound good to me.

---

Reviewed 5 of 12 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-Kcxm18t3567oQkJZmRJ:bijq7mr) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

it's a getter. 
</blockquote></details>

I meant getter, sorry. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcxnZMuREnH-7f0nQdV:boqhh9s)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
It would be good to create a Drake issue with the unfinished parts and this discussion about the pp Trajectory class, and then Russ might weigh in. 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcxoYw6WAD51zoK7-6t:b-vpa0r0)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (pending platform review).  Sorry I've been so slow at getting everything resolved.

---

Reviewed 5 of 12 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-Kcxq2N-ejcX17VlKhqa:b69pi1y) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

I meant getter, sorry. 
</blockquote></details>

BTW I'm OK with this,  I guess.  I keep seeing PRs where people want access to it for whatever reason.

---

*[drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcsEBX43HoJc4ZWVoan:-KcxqJ0m3sgFuyz3hp9_:b-oru1fo) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/examples/kuka_iiwa_arm/controlled_kuka/BUILD#L2)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

for cmake, i will just make sure that the shared stuff i added is updated accordingly. 
</blockquote></details>

works for me.

---

*[drake/systems/primitives/piecewise_polynomial_source.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcY__YIQf1aqifggSAw:-KcxqZgoSYpKq0-Zlika:boo8f4c) ([raw file](https://github.com/robotlocomotion/drake/blob/ab9b0adaffdf93d502505ae3c0a1bf83199af8b0/drake/systems/primitives/piecewise_polynomial_source.h#L1)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

i moved all that stuff to the actual demo code. i am leaving all the primitives as is. 
</blockquote></details>

Works for me.  Lucy or I can worry about making the functionality general once we need it (which might be soon in my case).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcxqYNFUqcQLX5NqBbn:bzensej)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-KcxsX1TvOPjDZJQ1PPn:be1phve) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW I'm OK with this,  I guess.  I keep seeing PRs where people want access to it for whatever reason.
</blockquote></details>

BTW ok, I'll release this complaint.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
I'll be creating an issue about adding derivative order to the pp Trajectory class. 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-Kcy4ERAhMkpU9wz9BNY:b-d059w1)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-Kcy7ivYaopfSDIgfpqY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@sammy-tri @david-german-tri, do you think it would be outrageously wrong for sammy to be both the platform and feature reviewer? it's a lot of stuff, and i hate to dump it on david... 

---

Review status: 34 of 44 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-Kcy9W48w0vUbLX66vvG:bocngf9)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-KcyA5Hg_mCy7PydpPw9:bapdw24) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
<details><summary><i>Previously, Lucy-tri (Lucy) wrote…</i></summary><blockquote>

BTW ok, I'll release this complaint.
</blockquote></details>

thanks guys. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 2 of 13 files at r8.
Review status: 36 of 44 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 1 of 13 files at r8.
Review status: 37 of 44 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
I'm on the fence about having only a single layer of review, though I'm OK with it if David is.

---

Reviewed 10 of 13 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-:-KcyW6WGzotQe8ZHU2Bc:b-6d2uw0)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Was there a review of the size of this PR?

Lucy-tri commented:
Reviewed 2 of 44 files at r1, 1 of 21 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/trajectories/piecewise_polynomial_trajectory.h, line 48 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4992#-KcxIlGwtiLiNcups4Mc:-KcyZPR6QS3i-yqH579I:byn9cwn) ([raw file](https://github.com/robotlocomotion/drake/blob/88433c46269b100160aa11f5107d78e8a45ebc70/drake/common/trajectories/piecewise_polynomial_trajectory.h#L48)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

thanks guys. 
</blockquote></details>

BTW this is being done in https://reviewable.io/reviews/RobotLocomotion/drake/4740#-, so watch out when you merge.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4992)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
@david-german-tri no. i wasn't aware of that rule actually. 
i will start chopping it up then. so should i just close this then?

david-german-tri commented:
> i will start chopping it up then. so should i just close this then?

Yeah, I think that's best.  Glancing through the PR, I also notice there are a bunch of new `.cc` files with no corresponding `_test.cc`.  As you PR this in smaller modules, please make sure those modules are adequately tested.
